lib/store_attribute/active_record/type/typed_store.rb

```diff
@@ -15,43 +15,57 @@ def self.create_from_type(basetype, key, type, **options)
 
       def initialize(subtype)
         @accessor_types = {}
+        @defaults = {}
         @store_accessor = subtype.accessor
         super(subtype)
       end
 
-      def add_typed_key(key, type, **options)
+      UNDEFINED = Object.new
+      private_constant :UNDEFINED
+
+      def add_typed_key(key, type, default: UNDEFINED, **options)
         type = ActiveRecord::Type.lookup(type, options) if type.is_a?(Symbol)
-        @accessor_types[key.to_s] = type
+        safe_key = key.to_s
+        @accessor_types[safe_key] = type
+        @defaults[safe_key] = default unless default == UNDEFINED
       end
 
       def deserialize(value)
         hash = super
-        if hash
-          accessor_types.each do |key, type|
-            hash[key] = type.deserialize(hash[key]) if hash.key?(key)
+        return hash unless hash
+        accessor_types.each do |key, type|
+          if hash.key?(key)
+            hash[key] = type.deserialize(hash[key])
+          elsif has_default?(key)
+            hash[key] = get_default(key)
           end
         end
         hash
       end
 
       def serialize(value)
-        if value.is_a?(Hash)
-          typed_casted = {}
-          accessor_types.each do |key, type|
-            k = key_to_cast(value, key)
-            typed_casted[k] = type.serialize(value[k]) unless k.nil?
+        return super(value) unless value.is_a?(Hash)
+        typed_casted = {}
+        accessor_types.each do |unsafe_key, type|
+          key = key_to_cast(value, unsafe_key)
+          next unless key
+          if value.key?(key)
+            typed_casted[key] = type.serialize(value[key])
+          elsif has_default?(key)
```

Comment:
Here we should use `unsafe_key`, 'cause `key` could be a symbol. Let's add a test scenario which reveals this problem.

And let's call it `str_key` instead (and also get rid of helper methods, e.g. `has_default?(key)` -> `defaults.key?(key)`.

Reply:
Is there any case code reaches this point? I haven't found one. What I'm saying is by removing `elsif` condition here I didn't see any changes, may be I'm missing something

Reply:
Removing this branch means the default will be not included to s11n. But we need it to match Attributes API behavior.

It actually makes sense to include the default to s11n, since the data may be deserialized elsewhere. For instance, on the client-side. And there is no guarantee the default will be available then.

I need to take take a closer look at this.

Reply:
Looks like `key_to_cast` needs to look up a current key in the `defaults` as well. Otherwise, the default value will not be included in s11n. Good find! I'll push an update a bit later today.